controller/lib/core/blower_fsm.h

```diff
@@ -155,11 +155,109 @@ class PressureControlFsm {
   Time expire_end_;
 };
 
+// "Breath finite state machine" for pressure assist mode.
+//
+// This mode is used for a patient that is not fully sedated and can initiate
+// breaths by himself, but once a breath has started, it is fully controlled
+// by the ventilator: inspiration time is set by the ventilator.
+// This mode enforce a minimum respiratory rate in case the patient does not
+// initiate enough breaths.
+// Similarly to pressure control, uses a square wave, only the length of the
+// exhale leg can vary depending on the patient breath initiation.
+// Unlike pressure control, this fsm starts with exhale and ends at the end of
+// inhale to be able to start the first inhale on patient trigger.
+class PressureAssistFsm {
+public:
+  explicit PressureAssistFsm(Time now, const VentParams &params)
+      : inspire_pressure_(cmH2O(static_cast<float>(params.pip_cm_h2o))),
+        expire_pressure_(cmH2O(static_cast<float>(params.peep_cm_h2o))),
+        start_time_(now),
+        latest_expire_end_(start_time_ + max_expire_duration(params)),
+        inspire_end_(latest_expire_end_ + inspire_duration(params)) {}
+
+  BlowerSystemState desired_state(Time now, const SensorReadings &readings) {
+
+    // TODO: change this when we get better understanding of inspiratory effort
```

Comment:
Could you extract this into a function that determines whether the patient is beginning an inhale?